cs/Markdown/TagsLibrary/MarkdownTagsLibraryFunctions.cs

```diff
@@ -0,0 +1,76 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Markdown
+{
+    public static partial class MarkdownTagsLibrary
+    {
+        public static bool TryToGetUsableTagInAssociations(string text, int startIndex, out TagElement resultTag)
+        {
+            if(text == null)
+                throw new NullReferenceException("text can't be null");
+
+            if(startIndex < 0 || startIndex >= text.Length)
+                throw new ArgumentOutOfRangeException("startIndex not included in the array bounds");
+
+            var parseVariants = new List<(TagElement tag, int priority)>();
+            var substrLength = 1;
+
+            while(substrLength <= MaximumTagLength && substrLength + startIndex <= text.Length)
+            {
+                var key = text.Substring(startIndex, substrLength);
+                if (TagAssociations.ContainsKey(key))
+                {
+                    var (type, priority) = TagAssociations[key];
+                    var tagUsability = GetTagUsability(text, startIndex, startIndex + substrLength - 1, type);
+                    
+                    if(tagUsability != TagUsability.None)
+                        parseVariants.Add((new TagElement(type, tagUsability, startIndex, substrLength), priority));
+                }
+                substrLength++;
+            }
+
+            if (parseVariants.Count == 0)
+            {
+                resultTag = null;
+                return false;
+            }
+
+            resultTag = parseVariants.OrderByDescending(x => x.priority).First().tag;
+            return true;
+        }
+
+        public static TagUsability GetTagUsability(string text, int startIndex, int endIndex, TagType type)
+        {
+            if (text == null)
+                throw new NullReferenceException();
+
+            if (startIndex < 0 || startIndex >= text.Length)
+                throw new ArgumentOutOfRangeException("startIndex not included in the array bounds");
+
+            if (endIndex < 0 || endIndex >= text.Length)
+                throw new ArgumentOutOfRangeException("endIndex not included in the array bounds");
```

Comment:
nit: 
```suggestion
                throw new ArgumentOutOfRangeException($"{nameof(endIndex)} is out of bounds");
```

Выше аналогично